cluster-autoscaler/core/scaledown/actuation/drain.go

```diff
@@ -19,128 +19,82 @@ package actuation
+				Err: errors.NewAutoscalerError(errors.UnexpectedScaleDownStateError, "Eviction did not attempted for the pod %s because some of the previous evictions failed", pod.Name)}
+		}
 	}
 
-	// Perform eviction of daemonset. We don't want to raise an error if daemonsetPod wasn't evict properly
-	for _, daemonSetPod := range daemonSetPods {
-		go func(podToEvict *apiv1.Pod) {
-			daemonSetConfirmations <- evictPod(ctx, podToEvict, true, retryUntil, e.EvictionRetryTime, e.evictionRegister)
-		}(daemonSetPod)
+	for _, group := range groups {
+		// If there are no pods in a particular range,
+		// then do not wait for pods in that priority range.
+		if len(group.Pods) == 0 {
+			continue
+		}
 
-	}
+		retryUntil := time.Now().Add(ctx.MaxPodEvictionTime)
+		maxEvictionDuration := ctx.MaxPodEvictionTime + 5*time.Second
 
-	podsEvictionCounter := 0
-	for i := 0; i < len(pods)+len(daemonSetPods); i++ {
-		select {
-		case evictionResult := <-confirmations:
-			podsEvictionCounter++
-			evictionResults[evictionResult.Pod.Name] = evictionResult
-			if evictionResult.WasEvictionSuccessful() {
-				metrics.RegisterEvictions(1)
-			}
-		case <-daemonSetConfirmations:
-		case <-time.After(retryUntil.Sub(time.Now()) + 5*time.Second):
-			if podsEvictionCounter < len(pods) {
-				// All pods initially had results with TimedOut set to true, so the ones that didn't receive an actual result are correctly marked as timed out.
-				return evictionResults, errors.NewAutoscalerError(errors.ApiCallError, "Failed to drain node %s/%s: timeout when waiting for creating evictions", node.Namespace, node.Name)
-			}
-			klog.Infof("Timeout when waiting for creating daemonSetPods eviction")
+		var err error
+		evictionResults, err = initiateEviction(ctx, node, group.Pods, evictionResults, retryUntil, maxEvictionDuration, evictionRetryTime, evictionRegister, group.ShutdownGracePeriodSeconds, ignoreDaemonSetPodEvictionResult)
+		if err != nil {
+			return evictionResults, err
 		}
-	}
 
-	evictionErrs := make([]error, 0)
-	for _, result := range evictionResults {
-		if !result.WasEvictionSuccessful() {
-			evictionErrs = append(evictionErrs, result.Err)
+		// Evictions created successfully, wait maxGracefulTerminationSec + podEvictionHeadroom to see if pods really disappeared.
+		evictionResults, err = waitPodsToDisappear(ctx, node, pods, evictionResults, podEvictionHeadroom, ignoreDaemonSetPodEvictionResult)
+		if err != nil {
+			return evictionResults, err
 		}
 	}
-	if len(evictionErrs) != 0 {
-		return evictionResults, errors.NewAutoscalerError(errors.ApiCallError, "Failed to drain node %s/%s, due to following errors: %v", node.Namespace, node.Name, evictionErrs)
-	}
+	return evictionResults, nil
+}
 
-	// Evictions created successfully, wait maxGracefulTerminationSec + podEvictionHeadroom to see if pods really disappeared.
+func waitPodsToDisappear(ctx *acontext.AutoscalingContext, node *apiv1.Node, pods []*apiv1.Pod, evictionResults map[string]status.PodEvictionResult,
```

Comment:
Same here - it's weird that a function called `waitForPodsToDisappear` only waits for some of the passed pods. Why not pass just the pods we do want to wait for (`fullEvictionPods`)?

Reply:
Done